Assignment6.java

```diff
@@ -31,36 +34,87 @@
             // If i isn't already empty.
             if (indexMap.get(0) != i) {
                 // Move the number in i to the empty slot so that i becomes empty.
-                Move emptyCurrentSlot = new Move(given[i], i, indexMap.get(0));
-                
-                listOfMoves.add(emptyCurrentSlot);
-                move(emptyCurrentSlot, given, indexMap);
+                listOfMoves.add(move(given[i], given, indexMap));
             }
 
             // Move the desired number to i.
-            Move moveToCurrentSlot = new Move(given[indexMap.get(desired[i])], indexMap.get(desired[i]), i);
-
-            listOfMoves.add(moveToCurrentSlot);
-            move(moveToCurrentSlot, given, indexMap);
+            listOfMoves.add(move(desired[i], given, indexMap));
         }
 
         return listOfMoves;
     }
+    
+    /**
+     * Rearranges given to make it the same as desired.
+     * Greedy algorithm to find the rearrangement with the least number of moves.
+     * O(n^2) worst case time complexity, O(n) space complexity.
+     *
+     * @param given   initial version of the array of numbers.
+     * @param desired final version of the array of numbers.
+     * @return list of moves made during arrangement.
+     */
+    public static List<Move> rearrangeGreedy(int[] given, int[] desired) {
+        List<Move> listOfMoves = new LinkedList<>();
+        
+        // Keys are values in given, values are their indexes in given.
+        Map<Integer, Integer> indexMap = populateIndexMap(given);
+
+        for (int car = nextCarToMove(given, desired, indexMap); car > 0;
+                car = nextCarToMove(given, desired, indexMap)) {
+            
+            listOfMoves.add(move(car, given, indexMap));
+        }
+        
+        return listOfMoves;
+    }
+    
+    /**
+     * Computes (greedily) the best choice of car to be moved next.
+     * O(n) worst case time complexity.
+     * 
+     * @param given    initial version of the array of numbers.
+     * @param desired  final version of the array of numbers.
+     * @param indexMap the map from cars to slots
+     * @return next car to be moved (0 in case of rearranged array)
+     */
+    private static int nextCarToMove(int[] given, int[] desired, Map<Integer, Integer> indexMap) {
+        //arrays of length 0
+        if (given.length == 0) {
+            return 0;
+        }
+        
+        //car belonging to the current empty slot
+        if (desired[indexMap.get(0)] != 0) {
+            return desired[indexMap.get(0)];
+        }
+        
+        //empty lot in its desired place => deblock the algorithm
+        //find the first car that isn't in its right place
+        for (int i = 0; i < given.length; i++) {

```

Comment:
If I understand correctly what you say, wouldn't we need some kind of global variable to know where to start the search next time?